### ISSUE TITLE
Proposal: throw error if no config file is found

### ISSUE THREAD

**Issue by nzakas** [eslint/eslint#5987]
From my travels around the internet, it seems that first-time ESLint users are very confused about the "nothing by default" mode. It seems many do not read the getting started docs, and so don't know to use `--init` to get started. That leaves them thinking that ESLint is broken because JSHint shows output out of the box. 

~~To remedy this, I'd like to propose that in 3.0, we change the default behavior to use `eslint:recommended ` when we don't find a config file. This should only affect CLI users and not affect `CLIEngine` by default (so as not to affect API consumers).~~

~~Another option is to automatically run `--init` when a config file isn't found, but that won't work well in editors, where an interactive prompt won't be available.~~

**Edit:** new proposal is throwing an error asking the user to run --init (we can use a pretty error message via messagsTemplate rather than showing the stack trace).

**Comment by platinumazure** [comment]
In order to avoid magic, could we just detect no configuration file and print a warning saying that user needs to create a config file?

This warning, plus maybe a CLI option (or `--init` sub-option) to auto-generate a minimal config that extends `eslint:recommended` would probably go a long way.

Another possibility, if we can reliably detect being in an interactive terminal or not, would be to magically run `--init` if no configuration only if we detect interactivity, otherwise output the warning I proposed in the first paragraph.

**Comment by nzakas** [comment]
I'm not sure there's much value in asking someone to use `--init` vs. just loading a reasonable config. If we are going through the trouble of detecting no config, it seems like we could be nice and give them something that works. Asking a new user to run another command when they just want to try ESLint just seems like another barrier to me (though admittedly is also better than what we have right now).

> Another possibility, if we can reliably detect being in an interactive terminal or not, would be to magically run --init if no configuration only if we detect interactivity, otherwise output the warning I proposed in the first paragraph.

👎 Way too magical for my tastes.

**Comment by platinumazure** [comment]
@nzakas Hmm. Well, I would say at the very least that we need to output a warning indicating any action we took, even if that is to create a minimal config that extends `eslint:recommended`. If we can agree to that stipulation, I can get behind your original proposal.

**Comment by nzakas** [comment]
@platinumazure I disagree that a warning is necessary. In this case, I think we are restoring expected functionality, so saying "Warning: ESLint is linting" does seem to add much value.

**Comment by alberto** [comment]
@nzakas I think it would be valuable to put a warning "no config file found, using eslint recommended configuration". It would be a little less magical and could help with human errors like a wrong config file name

**Comment by IanVS** [comment]
I think that since the recommended rules do not contain style rules (as our pre-1.0 defaults had), it would be perfectly fine to use `eslint:recommended` by default when there is no config file.  Seems like a very sensible default.  So, I'm :+1: to your proposal.

**Comment by mudetroit** [comment]
I don't think that most users would find it magical for a set of sane defaults to be run. I agree with @IanVS the default set should not contain style rules, but simply be focused around things that might cause your code to fail.

**Comment by ilyavolodin** [comment]
The whole reason why we removed defaults from ESLint was that users were complaining and not agreeing with our selection for default rules. Why would this time be any different then before? More than that, I feel this is going to create a lot of confusion, since right after installation ESLint will now check for 50-60 rules, but once you add config with a single rule, all of those 50-60 rules will magically disappear with no explanation (and as was mentioned before, users don't read documentation, so making a note about `extend: eslint:recommended` is not going to work either).
To summarize I'm very much against this proposal. I would, however be perfectly fine with just showing an error with instructions for `--init` if `.eslintrc` file is not found.

**Comment by nzakas** [comment]
@ilyavolodin this is different than what we had when people complained. The initial complaint was because we had a bunch of rules (both stylistic and not) that were enabled by default and then you'd need to manually go in and turn off each one you didn't want (merging your config with the default). That's why we introduced `--reset` and ultimately, decided that that should be the default. It was the pain of needing to turn off the things you didn't want that made people complain.

What I'm proposing here would apply only in cases where there are no other configs. Basically, right now we do this:

1. Look for local `.eslintrc` files and merge them
1. If there are no local `.eslintrc` files, then use `~/.eslintrc`
1. If there's no `~/.eslintrc`, then don't load anything.

And after the proposal, we'd do this:

1. Look for local `.eslintrc` files and merge them
1. If there are no local `.eslintrc` files, then use `~/.eslintrc`
1. If there's no `~/.eslintrc`, then use `eslint:recommended`

So we are not reintroducing the problem we had with v1.0, we're just providing a fallback for when there are no other configurations available.

I don't see this as being much different - indeed, for the majority of users, they would see no change in the behavior of ESLint. What I want to do is help the new users who might just want to explore ESLint without experimenting too much.

@alberto changing the output could affect downstream tooling, like editors. We can always output such a warning in the `--debug` log, I just don't think it makes sense in the regular output.

**Comment by ilyavolodin** [comment]
Here's a situation where I think this is going to create more problems then it will solve:

I'm a new user and I just installed ESLint and ran it on my repository without configuring it first. I got a bunch of errors and fixed them all. Now I continue to run ESLint with no config for a while on my repository and after a while I go to twitter and post "Love ESLint, but I wish it could check for indentations!". Somebody responds with "All you need to do is add `indent: 2` to your .eslintrc file and it will". So I go do that and get a bunch of indentation errors that I fix. Now for the next few weeks I keep running ESLint, but at this point what I don't realize is that the only rule that is checked by ESLint is `indent`. Since I fixed all of the previous errors, until I create a new error and remember that it should've been flagged, I have no way of knowing that default rules no longer apply.

This silent change is what I'm afraid of. I can see how this one change might result in a bunch of issues for us, create confusion for new users and in general make users very unhappy.

In my mind, exiting ESLint process with an error "No config file found, please run 'eslint --init` to configure. Exiting." is much more to the point and explain to users how to get started. If we do this with exit code 1, most of the editors should be able to pick up this error (since they have to be able to pick up parsing errors, which are display in the same format). This will stop CI builds, but that's probably a good thing as well.

**Comment by platinumazure** [comment]
I'm in agreement with @ilyavolodin-- this change would be *much worse* than what we have now. I'm now :-1: on anything more magical than "warn the user and exit if no config found".

**Comment by nzakas** [comment]
Agree. @ilyavolodin makes a strong case for why using a default config is a bad idea. Thanks for taking the time to explain (and for everyone watching, I'd hold this up as an example of how to argue your case, taking the time to write more than a couple sentences makes a huge difference).

I can get behind throwing an error asking the user to run `--init` (we can use a pretty error message via `messagsTemplate` rather than showing the stack trace).

Are people good with that?

**Comment by kaicataldo** [comment]
How do we handle working on an issue like this that we want for the next major version? Do we wait to make these changes until we have a better idea of when we'll be releasing the next major version, or are we free to work on these things now?

**Comment by ilyavolodin** [comment]
@kaicataldo If the change isn't breaking, we can work on it at any point. It's just marked as next milestone to be a prerequisite of the release of 3.0. If it's a breaking change, it's going to hang around unmerged until we create a branch and start preparing for 3.0 release. I think last comment from @nzakas makes this change non-breaking change, since we are not changing existing behavior, but extending it with clearer messaging.

**Comment by kaicataldo** [comment]
Makes sense - thanks for the thorough explanation, @ilyavolodin!

**Comment by nzakas** [comment]
@ilyavolodin this is still breaking. Right now, no config means you just parse the file, so it will still pick up parsing errors. With this change, it won't run at all.

It looks like we are getting a cluster of breaking proposals coming in, so we may need to do 3.0 soon to avoid a backlog. It would be nice to do a smaller major release than last time. :)

**Comment by ilyavolodin** [comment]
Sounds good. Agree, smaller release might be better, although I would want to hold off till we figure out JSCS compatibility issues. That way we can claim that 3.0 release is ready for JSCS migration.

**Comment by nzakas** [comment]
The JSCS compatibility tasks are not breaking, so likely wouldn't result in a major release anyway. I'd prefer not hold 3.0 for the sake of marketing and rather stick to semver where we do major releases only when breaking changes are introduced.

**Comment by zxqfox** [comment]
I'm not actually sure it's breaking. Who runs `eslint` in `nothing by default` mode? 😕

**Comment by ilyavolodin** [comment]
@zxqfox Brand new user that didn't read documentation?:-)

**Comment by nzakas** [comment]
If there's one thing we've learned -- there's no way to tell what people are actually doing with the software out in there in the world. :)

**Comment by zxqfox** [comment]
Oh, this is so true. ;-(

**Comment by platinumazure** [comment]
Anyone on this? If not, I'm happy to take a crack at it.

**Comment by platinumazure** [comment]
Oh, also, question on implementation: What is the expected behavior for the CLIEngine use case? Just throw? (I don't think the message template makes sense there because we can't assume the presence of a console, or am I wrong?)

**Comment by nzakas** [comment]
Hmm, that's a good question. I think `CLIEngine` would have to throw a regular error ("No ESLint configuration found.").

**Comment by platinumazure** [comment]
Sounds good to me, I'll get cracking on it this weekend or next week (or sooner if time allows).
